src/Documentation/Event-Sourcing/LogConsistencyProviders.md

```diff
@@ -0,0 +1,54 @@
+---
+layout: page
+title: Included Log-Consistency Providers
+---
+
+# Built-In Log-Consistency Providers
+
+The `Microsoft.Orleans.EventSourcing` package includes several log-consistency providers that cover basic scenarios suitable to get started, and allow some extensibility.
+
+
+### Orleans.EventSourcing.**StateStorage**.LogConsistencyProvider
+
+This provider stores *grain state snapshots*, using a standard storage provider that can be configured independently. 
+
+The data that is kept in storage is an object that contains both the grain state (as specified by the first type parameter to `JournaledGrain`) and some meta-data (the version number, amd a special tag that is used to avoid duplication of events when storage accesses fail).
+
+Since the entire grain state is read/written every time we access storage, this provider is not suitable for objects whose grain state is very large. 
+
+This provider does not support `RetrieveConfirmedEvents`: it cannot retrieve the events from storage because the events are not persisted.
```

Comment:
> the grain is maintaining some set of views that can change dynamically.

Typically replaying of the events does not happen at normal operation time. When you need to create a different view or process past events slightly differently, then that is typically done when the new version of the app is deployed. You won't go grain by grain as they get normally activated (you might though), but instead have a startup process that goes through the entire log and re-process it in some way... That's why you might need to tap into the specific storage semantics, instead of the narrow `ReadConfirmedEvents` that each grain would be able to access to once it gets activated.


> the grain wants to react to each event in some special way (in addition to just updating the grain state). For example, it may want to forward the event to some client.

Again, something that you can configure at the provider level to guarantee atomicity/transactions. Adding this method by itself can lead to hard to solve data consistency issues.

> About the snapshotting: we probably need a separate discussion on that, and write some higher-quality provider that actually does it. I never quite understood the details of how this is supposed to work. How does the provider know when it has to replay the log because the transitions have changed? After it fails to successfully deserialize the snapshot?

I guess this is partially answered by my first answer... this is not something that is typically done at runtime... You might have a deployment startup process that just deletes all the old snapshots for an entity and then starts replaying all events. We can help at the infrastructure level by adding the functions to each of the providers that allow snapshotting, by allowing reading the entire history, etc, but probably not by creating an abstraction for every provider, as they can all behave differently or be optimized for different needs, and hence the primary reason why people might choose to use one storage over the other. Also, it might be useful to build a mechanism to replay all events on the grains themselves, but at the same time be mindful that we are replaying the entire history so that we can optimize the access to storage. Nevertheless, I haven't thought about solutions to this in the context of Orleans yet. I'm only starting to look at what we have in our repo. Agreed that a separate discussion is needed